Manage conditions to hide or show questions in surveys

verarojman opened the pull request:
#### :tophat: What? Why?
Add new "logic" feature to questionnaires, allowing for questions to have **display conditions** to determine whether they should be shown, depending on the answer to another question from the same questionnaire.

**Condition types** are:
- _Answered_ or  _Not answered_ (They check whether the question has been answered or not)
- _Equal_ or _Not equal_ (They check whether the answer matches a given answer_option or not)
- _Includes text_ (It checks if answer contains given text, case-insensitively)

A condition can be **mandatory** or not. *All* mandatory conditions must be fulfilled for a question to be displayed. If no conditions are mandatory, the question will be displayed if *any* condition matches.

#### :pushpin: Related Issues
- Related to
  - [Metadecidim - Survey logic feature for Decidim Survey](https://meta.decidim.org/processes/roadmap/f/122/proposals/14719)
  - [OpenSourcePolitics - Survey logic feature for Decidim Survey](https://github.com/OpenSourcePolitics/dev/issues/3)

#### :clipboard: Subtasks
- [x] Add `CHANGELOG` entry
- [x] Add/modify seeds
- [x] Add tests

### :camera: Screenshots

#### Manage survey - Question with "includes text" condition
![Survey Logic - Condition type: Includes text](https://user-images.githubusercontent.com/8806781/75345617-5a850c00-589d-11ea-9883-95f2bbd683f2.png)
---

#### Manage survey - Question with multiple conditions
![Survey Logic - Multiple conditions](https://user-images.githubusercontent.com/8806781/75346328-c1ef8b80-589e-11ea-8abe-b22601cd4956.png)
---

#### Public behaviour video demo:
![Survey Logic - Public behaviour](https://user-images.githubusercontent.com/1401520/75538028-3d287d00-5a18-11ea-9b2e-475e07b405be.gif)

virgile-dev commented:
@microstudi could you set up a test instance so we can play with it ?

microstudi commented:
> @microstudi could you set up a test instance so we can play with it ? Thanks in advance !

sure, tomorrow I'll have one!

paulinebessoles commented:
Hi @microstudi and @verarojman ! 
Thank you for displaying the testing instance ! The feature is really cool, and more easy to use & understand than I thought 👍

I have one feedback about the "Answer option" drop-down field, in the display condition section. 
When the answer option is saved as "NO", the FO adopts the right behavior but in the survey editing page, the answer option drop-down still show the "YES" option. 
So if you save the form again, then the YES option is saved. 

To reproduce : 
1. Go to [http://decidim.platoniq.net/admin/participatory_processes/et-temporibus/components/14/manage/](http://decidim.platoniq.net/admin/participatory_processes/et-temporibus/components/14/manage/)
2. Change the display condition answer option for question #6 to NO
3. Save the survey
4. Go to question #6 and see the YES appearing in the drop-down field
5. Save and see FO survey 

<img width="1061" alt="Capture d’écran 2020-02-28 à 15 59 46" src="https://user-images.githubusercontent.com/52420208/75559289-7d512500-5a43-11ea-9bad-a84c89ff2cad.png">

verarojman commented:
Hi @paulinebessole !

Thank you for pointing that out, it should now work as expected once the last changes are integrated in the test instance (I cannot do that myself right now but I'll ask @microstudi).

> I have one feedback about the "Answer option" drop-down field, in the display condition section.
> When the answer option is saved as "NO", the FO adopts the right behavior but in the survey editing page, the answer option drop-down still show the "YES" option.
> So if you save the form again, then the YES option is saved.

microstudi commented:
@verarojman @paulinebessoles I've updated the instance! feel free to try again.
I've found somthing to correct  though, but it would be easy as it won't involve any javascript.

We've forgotten to take into account the view when people is not logged in the system, this is how it looks right know our demo:

![imatge](https://user-images.githubusercontent.com/1401520/75571184-f9099c80-5a58-11ea-8c19-bbafaf5a0273.png)

It should only display non-hidden options by default right? what do you think?

paulinebessoles commented:
> It should only display non-hidden options by default right? what do you think?

Thank you for the update ! You're right, I think the right behavior would be to only display questions without conditions.

microstudi commented:
@paulinebessoles @virgile-dev we forgot to take into account one case. What happens if the user has Javascript disabled? should we show the whole questionnaire? or just display a message to enable javascript to answer the survey? what do you think?

paulinebessoles commented:
Hi @microstudi , sorry for the delay of the answer! After discussing with our team tech, we'd say both of the answers : 

- If JS is disabled, a message should be displayed to warn the user of a reduced experience
- as it's a conditional survey, it should be okay to show all the questions in this case
Anyway, after the form has been submitted, a validation rule must not take into account the inconsistent responses 

What do you think ?

virgile-dev commented:
@verarojman Good job on this feature ! It's really great :)
I noticed a bug. 
As an admin when I create my set of question and didn't saved them yet if I click on add display condition and remove it as I have no question saved yet and try to save my survey the application crashes. When I try to refresh the page nothing has been saved and I have to start over creating my survey.

verarojman commented:
> @verarojman Good job on this feature ! It's really great :)
> I noticed a bug.
> As an admin when I create my set of question and didn't saved them yet if I click on add display condition and remove it as I have no question saved yet and try to save my survey the application crashes. When I try to refresh the page nothing has been saved and I have to start over creating my survey.

Thank you @virgile-dev! I'm working on the bug, will let you know when it's fixed.

microstudi commented:
I agree @verarojman , or maybe just left the button in a disabled state until the question is saved. In any case  we should avoid this kind of potential errors in the server level anyway.

verarojman commented:
> I agree @verarojman , or maybe just left the button in a disabled state until the question is saved. In any case we should avoid this kind of potential errors in the server level anyway.

Yes @microstudi , that's exactly what I thought. I'm making the button appear disabled until the question is saved, and also preventing the server-side error (it happened whenever a blank display condition was removed).

virgile-dev commented:
That's perfect for @verarojman let's go for the disabled button option.

tramuntanal commented:
can you check the conflicting files @verarojman please?

tramuntanal commented on decidim-forms/app/forms/decidim/forms/answer_form.rb:
```diff
@@ -47,8 +45,23 @@ def selected_choices
         choices.select(&:body)
       end
 
+      def conditions_fulfilled?
```
```suggestion
      def display_conditions_fulfilled?
```

tramuntanal commented on decidim-forms/app/models/decidim/forms/display_condition.rb:
```diff
@@ -0,0 +1,40 @@
+
+      belongs_to :question, class_name: "Question", foreign_key: "decidim_question_id", inverse_of: :display_conditions
+      belongs_to :condition_question, class_name: "Question", foreign_key: "decidim_condition_question_id", inverse_of: :display_conditions_for_other_questions
+      belongs_to :answer_option, class_name: "AnswerOption", foreign_key: "decidim_answer_option_id", optional: true
```
A display condition will always belong to a :question and sometimes also can belong to an :answer_option?
Also what's the role of a condition_question? should we document it here or in Question class?
Can we document it in the signature of this class?

tramuntanal commented on decidim-forms/app/models/decidim/forms/question.rb:
```diff
@@ -14,8 +14,28 @@ class Question < Forms::ApplicationRecord
+               dependent: :destroy,
+               inverse_of: :question
+
+      has_many :conditioned_questions,
```
Can you explain what a conditioned_question is please?

tramuntanal commented on decidim-forms/app/models/decidim/forms/question.rb:
```diff
@@ -14,8 +14,28 @@ class Question < Forms::ApplicationRecord
+               dependent: :destroy,
+               inverse_of: :question
+
+      has_many :display_conditions_for_other_questions,
```
So this associates display conditions to this question, but the name says "for other questions". Can you document what or which are "other questions" please?

tramuntanal commented on decidim-forms/app/presenters/decidim/forms/answer_option_presenter.rb:
```diff
@@ -0,0 +1,20 @@
+        @translated_body ||= translated_attribute body
+      end
+
+      def json
```
I would rather use the standards :to_json or :as_json
```suggestion
      def as_json
```

verarojman commented:
> Hi @verarojman ,
> Thanks for this feature!
> 
> I've started the review but I see it is complex. I'll run it locally to better understand everything. Do you mind applying the suggestions meanwhile and I'll come back with the review in a couple of days?
> 
> cc/ @microstudi

Thanks for the review, @tramuntanal! I've added some docstrings and fixed what you suggested, and I'll stay tuned for any feedback you have  :)

jesusdb commented:
Hi @verarojman, could you please check the conflicting files?

verarojman commented:
Is this (from `[CI] Admin / Tests`) a flaky test?

```
1st Try error in ./spec/system/admin_manages_newsletters_spec.rb:231:
Unable to find xpath "/html"
```

microstudi commented:
Yes, it's a flaky, it has been introduced recently. We'll try to solve it. Meanwhile I've restarted the admin test

jesusdb commented:
Hi @verarojman, could you please see the coverage of this method?
![imagen](https://user-images.githubusercontent.com/61017625/83251708-4ad1d800-a1aa-11ea-9548-e744eaf1e051.png)

verarojman commented:
@jesusdb this method is covered and I just added extra specs to ensure that. Are we okay to merge? :)

jesusdb commented:
Great! Now it seems to be covered, but it looks like there are others not covered? Could you please check the [answer_option.rb](https://codecov.io/gh/decidim/decidim/compare/2f70764380a6959d237c9478b5ad42171fe86b12...174a2b33c230c81347c43259cab06ac9061cafb3/src/decidim-forms/app/models/decidim/forms/answer_option.rb) and [update_questionnaire.rb](https://codecov.io/gh/decidim/decidim/compare/2f70764380a6959d237c9478b5ad42171fe86b12...174a2b33c230c81347c43259cab06ac9061cafb3/src/decidim-forms/app/commands/decidim/forms/admin/update_questionnaire.rb)?

microstudi commented:
@jesusdb these files should be covered now. Can you review this, thanks!

jesusdb commented on decidim-forms/app/views/decidim/forms/questionnaires/_answer.html.erb:
```diff
@@ -1,10 +1,9 @@
-<% field_id = "questionnaire_answers_#{answer_idx}" %>
+<% field_id = "questionnaire_responses_#{answer_idx}" %>
```
One doubt; why changing from "answers" to "responses"?

verarojman replied:
`questionnaire_responses_*` is the correct id, I merged from develop and had to change it.

microstudi replied:
@jesusdb this may be related to this: https://github.com/decidim/decidim/pull/5819 we changed answers to responses to avoid default behavior of Rectify Forms loading answers unnecessarily and causing a heavy penalty performance regression.

jesusdb commented on decidim-forms/spec/models/decidim/forms/display_condition_spec.rb:
```diff
@@ -0,0 +1,157 @@
+          expect(html_data[:condition]).to eq(subject.condition_question.id)
+        end
+
+        it "has an 'mandatory' attribute with the display_condition's 'mandatory' value" do
```
```suggestion
        it "has a 'mandatory' attribute with the display_condition's 'mandatory' value" do
```

verarojman commented on decidim-forms/lib/decidim/forms/test/shared_examples/manage_questionnaires/add_questions.rb:
```diff
@@ -0,0 +1,463 @@
+      end
+
+      page.all(".questionnaire-question").each_with_index do |question, idx|
+        question.all(".questionnaire-question-answer-option").each_with_index do |question_answer_option, aidx|
```
It meant "answer_option index", but I changed it to `answer_option_idx` just in case.

jesusdb commented:
@virgile-dev sorry for the delay! And thanks @verarojman, ready to merge!

PR superseded by #6241
